### ISSUE TITLE
Crash: Game crash on a single scenario when trying to build anything

### ISSUE THREAD

**Issue by Derpercoaster** [OpenLoco/OpenLoco#1274]
**Version information:**
 - Platform: Windows 10
 - Version: 21.10
 - Build/commit: 3bec790

**Describe the bug**
The game crashes every time I load a specific scenario and try to place anything on the map.

**To Reproduce**
Steps to reproduce the behaviour:
1. Load the scenario (provided below)
2. Just... try to do anything that involves construction

**Additional context**

Crash reports:

[5e19aac0-1765-4817-a2eb-6fb615e59757(OpenLoco, v21.10 (3bec790 on master)).zip](https://github.com/OpenLoco/OpenLoco/files/7645030/5e19aac0-1765-4817-a2eb-6fb615e59757.OpenLoco.v21.10.3bec790.on.master.zip)

Scenario:
[Austria1900.zip](https://github.com/OpenLoco/OpenLoco/files/7645035/Austria1900.zip)

**Comment by LeftofZen** [comment]
Wow this is crazy, can confirm this happens to be as well on v22.02

**Comment by LeftofZen** [comment]
A quick look in the debugger, the issue is deep in un-reverse-engineered code, but the caller appears to be `OpenLoco::MessageManager::post`

**Comment by LeftofZen** [comment]
Crash actually happens on 5th January (ie 5 days after opening scenario), regardless of whether you are building or not.
In MessageManager::post, these are the values being set before the crash:
![image](https://user-images.githubusercontent.com/7483209/153968301-2861c1e4-c775-4abe-932d-d35f2259a2e9.png)

**Comment by LeftofZen** [comment]
Setting the date past the 5th (ie to the 10th) and letting the game run then results in the same message crashing the game, just a few days later.
I think this is actually triggering on the update() call for the 6th, not that this changes anything. Callstack:
![image](https://user-images.githubusercontent.com/7483209/153968590-0f4f276a-2cea-4ecf-a2dd-e73dfb478dee.png)

**Comment by LeftofZen** [comment]
What is strange to me is that stations are attempting to be updated by StationManager, even though none exist at the start of the scenario. I tried building a station before the crash but that didn't solve anything.
I'll leave this for now, it requires a deeper investigation, but this is a strange bug indeed.

**Comment by LeftofZen** [comment]
One related bug I found is that the StationManager has an off-by-one bug in `id()` function:

Code is currently
` if (index > 1024)`

and it should be
` if (index >= Limits::maxStations)`

But I tested this fix and it didn't resolve this specific issue.

**Comment by LeftofZen** [comment]
I'll leave this for now, it requires a deeper investigation, but this is a strange bug indeed.

**Comment by LeftofZen** [comment]
Just thought to try it in vanilla and it crashes in the same fashion

**Comment by Derpercoaster** [comment]
> Just thought to try it in vanilla and it crashes in the same fashion

Which means the game isn't at fault so much as the scenario is. If it helps, I downloaded the scenario off of this website:
https://cache.locomotiondepot.net/

Which in turn I got to from here:
https://www.locomotiondepot.net/

The website is still being updated (at least according to the 2022 copyright), and it is https, so it seems safe enough. although I did run into a single malware-ish file there (unrelated to the Austria1900 file)

**Comment by Derpercoaster** [comment]
Also I must say LeftofZen, you jumped on quite a few of my posted issues rather quick. Thanks for that.

**Comment by LeftofZen** [comment]
> Which means the game isn't at fault so much as the scenario is.

Yeah it's an interesting one, I wonder whether the creator of the scenario ever tested their file by actually playing it. Maybe they did some weird hacking at the save with a memory editor or something, or maybe it just is a core bug in the game that this scenario exposes. I might play around with this more seriously on the weekend if I have some time as this bug is an interesting one.

> you jumped on quite a few of my posted issues rather quick

No problem, no-one likes making issues that appear to go into the void with no response :)

**Comment by Derpercoaster** [comment]
> maybe it just is a core bug in the game that this scenario exposes.

I was thinking something like "Maybe the specific layout of tiles, sprites etc. somehow causes the game to be unable to handle something." Games turn up weirdly specific issues sometimes. Of course, the original user using hacks isn't out of the question.

It might be helpful to try opening the save file in the scenario editor somehow. Generate a new landscape for it, save it as a playable game, and then see if the bug persists. If so, the bug is caused by the file itself. If not, then something in the specific scenario layout was at fault.

**Comment by LeftofZen** [comment]
> It might be helpful to try opening the save file in the scenario editor somehow. Generate a new landscape for it, save it as a playable game, and then see if the bug persists. If so, the bug is caused by the file itself. If not, then something in the specific scenario layout was at fault.

I just tried this - open up original scenario in scenario editor (no crashes, because `Station::update()` isn't called in editor mode), resaved the scenario as a new file, then opened that new file in a new game. Still crashes on the start of the 6th day.

I did manage to fix the original issue though, because for some reason there exists a valid station in memory in this scenario (so its a weird scenario file) even when there are no real stations in the game. The game crashes when trying to update this fake station, so I added a check if the station name has a blank station id which successfully bypasses the fake station. However around 1st Feb in the game, a new livestock fark begins construction and this again crashes the game in a different spot in the code, under `WindowManager::invalidateWidget()`. 

This leads me to believe there are issues with this scenario than just the name, making me uncertain that my quick fix is the right one.

**Comment by duncanspumpkin** [comment]
Are you sure the save wasnt created prior to https://github.com/OpenLoco/OpenLoco/pull/1070?

**Comment by LeftofZen** [comment]
I'm not sure when the scenario was created. If it really does have that problem as in #1070, is there a way to fix the scenario now?

**Comment by Derpercoaster** [comment]
> Are you sure the save wasnt created prior to #1070?

The save is stupidly old. It was created long before OpenLoco was a thing (not by me though). Although that doesn't explain why a "blank" station waspresent, since you can't exactly have stations in an unplayed scenario. Perhaps someone (ie the original creator of the scenario) messed with the code in an external program or something?

**Comment by LeftofZen** [comment]
Without having looked into this more (yet), and considering this crashes vanilla, my best guess is someone edited the .SC5 file in an editor of some kind, or included some kind of weird custom station object, or something along those lines.

**Comment by duncanspumpkin** [comment]
If its crashing vanilla then I don't see there being any worth investigating further for now.

**Comment by Derpercoaster** [comment]
> If its crashing vanilla then I don't see there being any worth investigating further for now.

Which it does.

**Comment by Derpercoaster** [comment]
> Even the MoneyEffects glitch out, being shown in the wrong colour (in the company colour actually)

I know this is an old comment, but the money effects are always in company colors, aren't they?

**Comment by duncanspumpkin** [comment]
> > Even the MoneyEffects glitch out, being shown in the wrong colour (in the company colour actually)
> 
> I know this is an old comment, but the money effects are always in company colors, aren't they?

Technically there are two types of money effect company colour ones and legacy rct2 ones. You can't make rct2 ones but if you corrupted your save you might be able to make them.

**Comment by Derpercoaster** [comment]
Ah, I see. I guess it makes sense since so much code from RCT2 was recycled into Locomotion.
